pkg/registry/thirdpartyresourcedata/codec.go

```diff
@@ -0,0 +1,172 @@
+	return mapping, nil
+}
+
+func (t *thirdPartyResourceDataMapper) AliasesForResource(resource string) ([]string, bool) {
+	return t.mapper.AliasesForResource(resource)
+}
+
+func (t *thirdPartyResourceDataMapper) ResourceSingularizer(resource string) (singular string, err error) {
+	return t.mapper.ResourceSingularizer(resource)
+}
+
+func (t *thirdPartyResourceDataMapper) VersionAndKindForResource(resource string) (defaultVersion, kind string, err error) {
+	return t.mapper.VersionAndKindForResource(resource)
+}
+
+func NewMapper(mapper meta.RESTMapper, kind string) meta.RESTMapper {
+	return &thirdPartyResourceDataMapper{mapper, kind}
+}
+
+type thirdPartyResourceDataCodec struct {
+	delegate runtime.Codec
+	kind     string
+}
+
+func NewCodec(codec runtime.Codec, kind string) runtime.Codec {
+	return &thirdPartyResourceDataCodec{codec, kind}
+}
+
+func (t *thirdPartyResourceDataCodec) populate(objIn *expapi.ThirdPartyResourceData, data []byte) error {
+	var obj interface{}
+	if err := json.Unmarshal(data, &obj); err != nil {
+		fmt.Printf("Invalid JSON:\n%s\n", string(data))
+		return err
+	}
+	mapObj, ok := obj.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("unexpected object: %#v", obj)
+	}
+	kind, ok := mapObj["kind"].(string)
+	if !ok {
+		return fmt.Errorf("unexpected object: %#v", obj)
+	}
+	if kind != t.kind {
+		return fmt.Errorf("unexpected kind: %s, expected: %s", kind, t.kind)
+	}
+
+	metadata, ok := mapObj["metadata"].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("unexpected object: %#v", obj)
+	}
+
+	name, ok := metadata["name"].(string)
+	if !ok {
+		return fmt.Errorf("unexpected object: %#v", obj)
+	}
+
+	objIn.Name = name
+	objIn.Data = data
+	return nil
+}
+
+func (t *thirdPartyResourceDataCodec) Decode(data []byte) (runtime.Object, error) {
+	result := &expapi.ThirdPartyResourceData{}
+	if err := t.populate(result, data); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (t *thirdPartyResourceDataCodec) DecodeToVersion(data []byte, version string) (runtime.Object, error) {
+	// TODO: this is hacky, there must be a better way...
+	obj, err := t.Decode(data)
+	if err != nil {
+		return nil, err
+	}
+	objData, err := t.delegate.Encode(obj)
+	if err != nil {
+		return nil, err
+	}
+	return t.delegate.DecodeToVersion(objData, version)
+}
+
+func (t *thirdPartyResourceDataCodec) DecodeInto(data []byte, obj runtime.Object) error {
+	thirdParty, ok := obj.(*expapi.ThirdPartyResourceData)
+	if !ok {
+		return fmt.Errorf("unexpected object: %#v", obj)
+	}
+	return t.populate(thirdParty, data)
+}
+
+func (t *thirdPartyResourceDataCodec) DecodeIntoWithSpecifiedVersionKind(data []byte, obj runtime.Object, kind, version string) error {
+	thirdParty, ok := obj.(*expapi.ThirdPartyResourceData)
+	if !ok {
+		return fmt.Errorf("unexpected object: %#v", obj)
+	}
+	if err := t.populate(thirdParty, data); err != nil {
+		return err
+	}
+	thirdParty.Kind = kind
+	thirdParty.APIVersion = version
```

Comment:
Also, DecodeIntoWithSpecifiedVersionKind() does two things: 

* check if data.APIVersion and data.Kind matched with the specified ones (which are populated from reqScope and are supposed to be identical to the version and kind in the URL); 

* if data.APIVersion and data.Kind are empty, they will be default to the specified ones. Defaulting is useful when user does a POST or PUT without specifying the Kind or the Version in the json/yaml config file. 

The specified kind is always ThirdPartyResourceData, which won't match data.Kind anyway, so we can omit that check. Could we at least check and default data.Version? And perhaps we should comment here that checking and defaulting of data.Kind is not implemented.